src/ng/compile.js

```diff
@@ -2540,9 +2545,23 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
         lastValue,
         parentGet, parentSet, compare;
 
+        if ((attrName in attrs) && !attrs.hasOwnProperty(attrName)) {
+          // User defined a binding with same as a method in Object.prototype and also the attribute is not present
+          // Override property to make sure subsequent code do not have access to prototype method
+          attrs[attrName] = undefined;
+        }
+
         switch (mode) {
 
           case '@':
+            if (!attrs[attrName]) {
+              if (optional) {
+                break;
```

Comment:
so if the value is optional and it does not have an initial value, then we are not observing the attribute? This implies that
`<div attr-abc="{{abc}}" some-directive-that-binds-to-the-attribute-abc></div>`
will stop working. This is a breaking change